mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -313,133 +313,139 @@ class Word2Vec extends Serializable with Logging {
                   }
+                  a += 1
                 }
-                a += 1
+                pos += 1
               }
-              pos += 1
+              (syn0, syn1, lwc, wc)
+          }
+          val syn0Local = model._1
+          val syn1Local = model._2
+          // Only output modified vectors.
+          Iterator.tabulate(vocabSize) { index =>
+            if (syn0Modify(index) > 0) {
+              Some((index, syn0Local.slice(index * vectorSize, (index + 1) * vectorSize)))
+            } else {
+              None
+            }
+          }.flatten ++ Iterator.tabulate(vocabSize) { index =>
+            if (syn1Modify(index) > 0) {
+              Some((index + vocabSize, syn1Local.slice(index * vectorSize, (index + 1) * vectorSize)))
+            } else {
+              None
             }
-            (syn0, syn1, lwc, wc)
+          }.flatten
         }
-        val syn0Local = model._1
-        val syn1Local = model._2
-        // Only output modified vectors.
-        Iterator.tabulate(vocabSize) { index =>
-          if (syn0Modify(index) > 0) {
-            Some((index, syn0Local.slice(index * vectorSize, (index + 1) * vectorSize)))
-          } else {
-            None
-          }
-        }.flatten ++ Iterator.tabulate(vocabSize) { index =>
-          if (syn1Modify(index) > 0) {
-            Some((index + vocabSize, syn1Local.slice(index * vectorSize, (index + 1) * vectorSize)))
-          } else {
-            None
-          }
-        }.flatten
-      }
-      val synAgg = partial.reduceByKey { case (v1, v2) =>
+        val synAgg = partial.reduceByKey { case (v1, v2) =>
           blas.saxpy(vectorSize, 1.0f, v2, 1, v1, 1)
           v1
-      }.collect()
-      var i = 0
-      while (i < synAgg.length) {
-        val index = synAgg(i)._1
-        if (index < vocabSize) {
-          Array.copy(synAgg(i)._2, 0, syn0Global, index * vectorSize, vectorSize)
-        } else {
-          Array.copy(synAgg(i)._2, 0, syn1Global, (index - vocabSize) * vectorSize, vectorSize)
+        }.collect()
+        var i = 0
+        while (i < synAgg.length) {
+          val index = synAgg(i)._1
+          if (index < vocabSize) {
+            Array.copy(synAgg(i)._2, 0, syn0Global, index * vectorSize, vectorSize)
+          } else {
+            Array.copy(synAgg(i)._2, 0, syn1Global, (index - vocabSize) * vectorSize, vectorSize)
+          }
+          i += 1
         }
-        i += 1
+        bcSyn0Global.unpersist(false)
+        bcSyn1Global.unpersist(false)
       }
-      bcSyn0Global.unpersist(false)
-      bcSyn1Global.unpersist(false)
-    }
-    newSentences.unpersist()
-    expTable.destroy()
-    bcVocab.destroy()
-    bcVocabHash.destroy()
+      newSentences.unpersist()
 
-    val wordArray = vocab.map(_.word)
-    new Word2VecModel(wordArray.zipWithIndex.toMap, syn0Global)
+      val wordArray = vocab.map(_.word)
+      new Word2VecModel(wordArray.zipWithIndex.toMap, syn0Global)
+    }
+    finally
```

Comment:
btw. are you aare of http://jsuereth.com/scala-arm/ which might prove damn useful to ease resource management without clustering (much) the syntax. It is akin to using clauses in C# or context managers in Python.